TypeSystem/variant.h

```diff
@@ -259,52 +191,69 @@ struct VariantImpl<NAME, TypeListImpl<TYPES...>> : CurrentVariantImpl<NAME> {
   }
 
   template <typename X>
-  ENABLE_IF<std::is_same<X, VariantImpl>::value, const VariantImpl&> VariantValueImpl() const {
+  std::enable_if_t<std::is_same<X, VariantImpl>::value, const VariantImpl&> VariantValueImpl() const {
     if (ExistsImpl()) {
       return *this;
     } else {
       throw NoValueOfTypeException<VariantImpl>();
     }
   }
 
-  void CheckIntegrityImpl() const {
-    // NOTE by @dkorolev on 5/1/16: Should we be OK to have an uninitialized `Variant<>`?
-    // TODO(dkorolev) + TODO(mzhurovich): Revisit this.
-    if (!object_) {
-      throw UninitializedVariantOfTypeException<TYPES...>();  // LCOV_EXCL_LINE
-    }
-  }
-
  private:
   struct TypeAwareClone {
-    variant_impl_t& result;
-    TypeAwareClone(variant_impl_t& result) : result(result) {}
+    std::unique_ptr<CurrentSuper>& destination;
+    TypeAwareClone(std::unique_ptr<CurrentSuper>& destination) : destination(destination) {}
+
+    template <typename U>
+    std::enable_if_t<TypeListContains<typelist_t, current::decay<U>>::value> operator()(const U& instance) {
+      destination = std::make_unique<current::decay<U>>(instance);
+    }
 
-    template <typename TT>
-    void operator()(const TT& instance) {
-      result = instance;
+    template <typename U>
+    std::enable_if_t<!TypeListContains<typelist_t, U>::value> operator()(const U&) {
```

Comment:
Inconsistency w/ `decay<>` present in the next line but not present in this one.